internal/daemon/ucrednet.go

```diff
@@ -109,18 +109,21 @@ func (wl *ucrednetListener) Accept() (net.Conn, error) {
 
 	var unet *ucrednet
 	if ucon, ok := con.(*net.UnixConn); ok {
-		f, err := ucon.File()
+		rawConn, err := ucon.SyscallConn()
 		if err != nil {
 			return nil, err
 		}
-		// File() is a dup(); needs closing
-		defer f.Close()
-
-		ucred, err := getUcred(int(f.Fd()), syscall.SOL_SOCKET, syscall.SO_PEERCRED)
+		var ucred *syscall.Ucred
+		var ucredErr error
+		err = rawConn.Control(func(fd uintptr) {
+			ucred, ucredErr = getUcred(int(fd), syscall.SOL_SOCKET, syscall.SO_PEERCRED)
+		})
```

Comment:
Yeah, this took us a while to figure out. It's because the code (inherited from snapd, see [here](https://github.com/snapcore/snapd/blob/a79e60b1cb335d7588ca1973f928751498c01ca4/daemon/ucrednet.go#L126)) used `ucon.File()` to get the fd, causing a race condition in the credentials-checking code that meant websocket connections failed. The lxd code (see [here](https://github.com/lxc/lxd/blob/1131deba68b8672ea2769c7dfd474cb04d1e359c/lxd/ucred/ucred.go#L19)) for the same thing uses `conn.Control`, which ensures the fd is valid for the duration of the call. I've added a brief comment explaining.

I was sure I opened a PR against snapd, but I guess not -- can't for the life of me find it. In any case, I guess they don't have this issue (no websockets in snapd?).

Reply:
Yeah, no websockets there for now. Thanks for explaining and for commenting the code.